packages/infinite-scroll/src/index.ts

```diff
@@ -0,0 +1,141 @@
+export const DEFAULT_DELAY = 200
+export const DEFAULT_DISTANCE = 0
+
+const attributes = {
+  delay: {
+    type: Number,
+    default: DEFAULT_DELAY,
+  },
+  distance: {
+    type: Number,
+    default: DEFAULT_DISTANCE,
+  },
+  disabled: {
+    type: Boolean,
+    default: false,
+  },
+  immediate: {
+    type: Boolean,
+    default: true,
+  },
+}
+
+type Attrs = typeof attributes
+type ScrollOptions = { [K in keyof Attrs]: Attrs[K]['default'] }
+type InfiniteScrollCallback = () => void
+type InfiniteScrollEl = HTMLElement & {
+  [SCOPE]: {
+    container: HTMLElement | Window
+    containerEl: HTMLElement
+    instance: ComponentPublicInstance
+    cb: InfiniteScrollCallback
+    onScroll: () => void
+    observer?: MutationObserver
+  }
+}
+
+const getScrollOptions = (el: HTMLElement, instance: ComponentPublicInstance): ScrollOptions => {
+  return entries(attributes)
+    .reduce((acm, [name, option]) => {
+      const { type, default: defaultValue } = option
+      const attrVal = el.getAttribute(`infinite-scroll-${name}`)
+      let value = instance[attrVal] ?? attrVal ?? defaultValue
+      value = value === 'false' ? false : value
+      value = type(value)
+      acm[name] = isNaN(value) ? defaultValue : value
+      return acm
+    }, {} as ScrollOptions)
+}
+
+const destroyObserver = (el: InfiniteScrollEl) => {
+  const { observer } = el[SCOPE]
+
+  if (observer) {
+    observer.disconnect()
+    delete el[SCOPE].observer
+  }
+}
+
+const handleScroll = (el: InfiniteScrollEl, cb: InfiniteScrollCallback) => {
+  const { container, containerEl, instance, observer } = el[SCOPE]
+  const { disabled, distance } = getScrollOptions(el, instance)
+
+  // trigger only if full check has done and not disabled
+  if (observer || disabled) return
+
+  const { clientHeight, scrollHeight, scrollTop } = containerEl
+  let shouldTrigger = false
+
+  if (container === el) {
+    shouldTrigger = scrollHeight - (clientHeight + scrollTop) <= distance
+  } else {
+    // get the scrollHeight since el might be visible overflow
+    const { clientTop, scrollHeight: height } = el
+    const offsetTop = getOffsetTopDistance(el, containerEl)
+    shouldTrigger = scrollTop + clientHeight >= offsetTop + clientTop + height - distance
+  }
+
+  if (shouldTrigger) {
+    cb.call(instance)
+  }
+}
+
+function checkFull(el: InfiniteScrollEl, cb: InfiniteScrollCallback) {
+  const { containerEl, instance } = el[SCOPE]
+  const { disabled } = getScrollOptions(el, instance)
+
+  if (disabled) return
+
+  if (containerEl.scrollHeight <= containerEl.clientHeight) {
+    cb.call(instance)
+  } else {
+    destroyObserver(el)
+  }
+}
+
+const InfiniteScroll: ObjectDirective<InfiniteScrollEl, InfiniteScrollCallback> = {
+  async mounted(el, binding) {
+    const { instance, value: cb } = binding
+
+    if (!isFunction(cb)) {
+      throwError(SCOPE, '\'v-infinite-scroll\' binding value must be a function')
+    }
+
+    // ensure parentNode mounted
+    await nextTick()
```

Comment:
I think at the time which `directive` gets attached, the DOM should already be mounted, I don't think you need to do the extra `nextTick` right here. Refer to https://v3.vuejs.org/guide/migration/custom-directives.html#_3-x-syntax, you could see that according the documentation, by the time `beforeMounted` gets invoked, `el` has already been mounted on DOM tree. have you encountered any problems here?

Reply:
Truely, when `mounted` gets invoked, the current DOM has been mounted. But, there is no guarantee that all the parent nodes have been mounted. So I have to wait a moment until all the parent nodes are mounted in order to get the scroll container which might not included in el. I haven't read the source code and not sure its reliability, but it does work in my case.

Reply:
Well, this is my mistake, the way how you handling this is correct.